### PR TITLE
get logs for services deployed in kubernetes and docker

### PR DESCRIPTION
**Pull request by khannaabhi**
1) added getLogs to interface and routes
2)  getlog function in docker and istio get container logs and the sends data line by line
3) as gateway make request to runner, so i implemented code to read chunked data if proper header in passed from runner

### CODE REVIEW

**Comment by YourTechBud** [comment]
@khannaabhi resolve conflicts

File: gateway/go.mod [hunk 1]
```diff
@@ -18,6 +18,7 @@ require (
 	github.com/ghodss/yaml v1.0.0
 	github.com/go-sql-driver/mysql v1.5.0
 	github.com/gogo/protobuf v1.3.1 // indirect
+	github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b
```
Comment (YourTechBud) [inline gateway/go.mod:21]: Run `go mod tidy`

File: gateway/managers/syncman/syncman_runner.go [hunk 2]
```diff
@@ -288,6 +291,107 @@ func (s *Manager) HandleRunnerGetServiceRoutingRequest(admin *admin.Manager) htt
...
+		// Copy headers and status code
+		for k, v := range response.Header {
+			// check if data is available in chunks
+			if k == "X-Content-Type-Options" && v[0] == "nosniff" {
+				streamData = true
+			}
+			w.Header().Set(k, v[0])
+		}
+
+		if streamData {
```
Comment (YourTechBud) [inline gateway/managers/syncman/syncman_runner.go:354]: You dont seem to send any response if streamData is false

File: gateway/server/routes.go [hunk 3]
```diff
@@ -159,6 +159,7 @@ func (s *Server) routes(profiler bool, staticPath string, restrictedHosts []stri
 	runnerRouter.Methods(http.MethodDelete).Path("/{project}/services/{serviceId}/{version}").HandlerFunc(s.managers.Sync().HandleRunnerDeleteService(s.managers.Admin()))
 	runnerRouter.Methods(http.MethodPost).Path("/{project}/service-routes/{serviceId}").HandlerFunc(s.managers.Sync().HandleRunnerServiceRoutingRequest(s.managers.Admin()))
 	runnerRouter.Methods(http.MethodGet).Path("/{project}/service-routes").HandlerFunc(s.managers.Sync().HandleRunnerGetServiceRoutingRequest(s.managers.Admin()))
+	runnerRouter.Methods(http.MethodGet).Path("/{project}/services/logs/{taskId}/{replicaId}").HandlerFunc(s.managers.Sync().HandleRunnerGetServiceLogs(s.managers.Admin()))
```
Comment (YourTechBud) [inline gateway/server/routes.go:162]: I thought taskId was optional. Is it not?

File: runner/server/handle.go [hunk 4]
```diff
@@ -135,6 +136,41 @@ func (s *Server) handleApplyService() http.HandlerFunc {
...
+		isFollow, err := strconv.ParseBool(follow[0])
+		if err != nil {
+			logrus.Errorf("Failed to get service logs - %s", err.Error())
+			_ = utils.SendErrorResponse(w, http.StatusInternalServerError, err.Error())
+			return
+		}
+
+		utils.LogDebug("Get logs process started", "docker", "GetLogs", map[string]interface{}{"projectId": projectID, "taskId": taskID, "replicaId": replicaID, "isFollow": isFollow})
+
+		if err := s.driver.GetLogs(ctx, isFollow, projectID, taskID, replicaID, w, r); err != nil {
```
Comment (YourTechBud) [inline runner/server/handle.go:166]: I think `GetLogs` should return a reader. Internally you can use io.Pipe to achieve this.

File: runner/utils/driver/istio/apply.go [hunk 5]
```diff
@@ -131,3 +140,58 @@ func (i *Istio) ApplyServiceRoutes(ctx context.Context, projectID, serviceID str
 
 	return i.applyVirtualService(ctx, ns, virtualService)
 }
+
+// GetLogs get logs of specified services
+func (i *Istio) GetLogs(ctx context.Context, isFollow bool, projectID, taskID, replica string, w http.ResponseWriter, r *http.Request) error {
```
Comment (YourTechBud) [inline runner/utils/driver/istio/apply.go:145]: This should be in a file `logs.go`

File: space-cli/cmd/cmd.go [hunk 6]
```diff
@@ -2,6 +2,7 @@ package cmd
 
 import (
 	"fmt"
+	"github.com/spaceuptech/space-cloud/space-cli/cmd/modules/logs"
```
Comment (YourTechBud) [inline space-cli/cmd/cmd.go:5]: Not ordered right

File: space-cli/cmd/modules/getter.go [hunk 7]
```diff
@@ -29,6 +30,7 @@ func FetchGetSubCommands() *cobra.Command {
 	getCmd.AddCommand(project.GetSubCommands()...)
 	getCmd.AddCommand(remoteservices.GetSubCommands()...)
 	getCmd.AddCommand(services.GetSubCommands()...)
+	getCmd.AddCommand(logs.GetSubCommands()...)
```
Comment (YourTechBud) [inline space-cli/cmd/modules/getter.go:33]: This shouldn't be here

File: space-cli/cmd/modules/operations/setup.go [hunk 8]
```diff
@@ -205,6 +205,7 @@ func Setup(username, key, config, version, secret, clusterName string, dev bool,
 				"ROUTING_FILE_PATH=" + "/routing-config.json",
 				"CLUSTER_ID=" + clusterID,
 				"PORT=4050",
+				"LOG_LEVEL=debug",
```
Comment (YourTechBud) [inline space-cli/cmd/modules/operations/setup.go:208]: Remove this. Its taken care of in another PR

File: gateway/managers/syncman/syncman_runner.go (hunk 2 above)
Comment (sharadregoti) [reply gateway/managers/syncman/syncman_runner.go:354]: should i send error response specifying you need to set this header

File: gateway/server/routes.go (hunk 3 above)
Comment (sharadregoti) [reply gateway/server/routes.go:162]: ok, so should I take it from query param?

File: gateway/managers/syncman/syncman_runner.go (hunk 2 above)
Comment (YourTechBud) [reply gateway/managers/syncman/syncman_runner.go:354]: Ok. Thats works

File: gateway/server/routes.go (hunk 3 above)
Comment (YourTechBud) [reply gateway/server/routes.go:162]: Yes. That would be better. Even replica id could be in query param.

File: gateway/managers/syncman/syncman_runner.go [hunk 9]
```diff
@@ -288,6 +292,110 @@ func (s *Manager) HandleRunnerGetServiceRoutingRequest(admin *admin.Manager) htt
...
+			}
+			w.Header().Set("X-Content-Type-Options", "nosniff")
+
+			for {
+				select {
+				case <-done:
+					utils.LogDebug("Client stopped listening", "syncman", "HandleRunnerGetServiceLogs", nil)
+					return
+				default:
+					str, _ := rd.ReadString('\n')
```
Comment (YourTechBud) [inline gateway/managers/syncman/syncman_runner.go:385]: Why aren't you checking for an error here?

File: runner/go.mod [hunk 10]
```diff
@@ -28,6 +28,7 @@ require (
 	github.com/txn2/txeh v1.3.0
 	github.com/urfave/cli v1.22.2
 	go.etcd.io/bbolt v1.3.3
+	go.uber.org/zap v1.15.0
```
Comment (YourTechBud) [inline runner/go.mod:31]: run `go mod tidy`

File: runner/server/handle.go [hunk 11]
```diff
@@ -135,6 +136,70 @@ func (s *Server) handleApplyService() http.HandlerFunc {
...
+		reader := bufio.NewReader(pipeReader)
+		// implement http flusher
+		flusher, ok := w.(http.Flusher)
+		if !ok {
+			_ = utils.SendErrorResponse(w, http.StatusBadRequest, "expected http.ResponseWriter to be an http.Flusher")
+			return
+		}
+		w.Header().Set("X-Content-Type-Options", "nosniff")
+		w.WriteHeader(http.StatusOK)
+		utils.LogDebug("before for loop", "docker", "GetLogs", map[string]interface{}{"projectId": projectID, "taskId": taskID, "replicaId": replicaID, "isFollow": isFollow})
```
Comment (YourTechBud) [inline runner/server/handle.go:175]: Remove this

File: runner/server/handle.go [hunk 12]
```diff
@@ -135,6 +136,70 @@ func (s *Server) handleApplyService() http.HandlerFunc {
...
+
+		taskID := r.URL.Query().Get("taskId")
+		replicaID := r.URL.Query().Get("replicaId")
+		if replicaID == "" {
+			_ = utils.SendErrorResponse(w, http.StatusInternalServerError, "replica id not provided in query param")
+			return
+		}
+		_, isFollow := r.URL.Query()["follow"]
+
+		utils.LogDebug("Get logs process started", "docker", "GetLogs", map[string]interface{}{"projectId": projectID, "taskId": taskID, "replicaId": replicaID, "isFollow": isFollow})
```
Comment (YourTechBud) [inline runner/server/handle.go:157]:
```suggestion
		utils.LogDebug("Get logs process started", "handler", "get-logs", map[string]interface{}{"projectId": projectID, "taskId": taskID, "replicaId": replicaID, "isFollow": isFollow})
```

File: runner/utils/driver/docker/docker.go [hunk 13]
```diff
@@ -124,6 +126,68 @@ func (d *Docker) ApplyService(ctx context.Context, service *model.Service) error
...
+	}
+
+	pipeReader, pipeWriter := io.Pipe()
+	utils.LogDebug("Sending logs to client", "docker", "GetLogs", map[string]interface{}{})
+	go func() {
+		defer utils.CloseTheCloser(b)
+		defer utils.CloseTheCloser(pipeWriter)
+		// Read logs
+		rd := bufio.NewReader(b)
+		utils.LogDebug("go fund", "", "", nil)
```
Comment (YourTechBud) [inline runner/utils/driver/docker/docker.go:171]: remove this

File: runner/utils/driver/docker/docker.go [hunk 14]
```diff
@@ -124,6 +126,68 @@ func (d *Docker) ApplyService(ctx context.Context, service *model.Service) error
...
+	if err != nil {
+		logrus.Errorf("unable to list containers got error message - %v", err)
+		return nil, err
+	}
+	var b io.ReadCloser
+	containerNotFound := true
+	for _, container := range containers {
+		if strings.HasSuffix(container.Names[0], taskID) {
+			containerNotFound = false
+			utils.LogDebug("Requesting logs from docker client", "docker", "GetLogs", map[string]interface{}{"containerName": container.Names, "isFollow": isFollow})
```
Comment (YourTechBud) [inline runner/utils/driver/docker/docker.go:151]:
```suggestion
			utils.LogDebug("Requesting logs from docker client", "docker", "get-logs", map[string]interface{}{"containerName": container.Names, "isFollow": isFollow})
```

File: runner/utils/driver/docker/docker.go [hunk 15]
```diff
@@ -124,6 +126,68 @@ func (d *Docker) ApplyService(ctx context.Context, service *model.Service) error
...
+			break
+		}
+	}
+
+	if containerNotFound {
+		return nil, fmt.Errorf("Unable to find specified container, check if the container is running")
+	}
+
+	pipeReader, pipeWriter := io.Pipe()
+	utils.LogDebug("Sending logs to client", "docker", "GetLogs", map[string]interface{}{})
```
Comment (YourTechBud) [inline runner/utils/driver/docker/docker.go:165]:
```suggestion
	utils.LogDebug("Sending logs to client", "docker", "get-logs", map[string]interface{}{})
```

File: runner/utils/driver/docker/docker.go [hunk 16]
```diff
@@ -124,6 +126,68 @@ func (d *Docker) ApplyService(ctx context.Context, service *model.Service) error
...
+	go func() {
+		defer utils.CloseTheCloser(b)
+		defer utils.CloseTheCloser(pipeWriter)
+		// Read logs
+		rd := bufio.NewReader(b)
+		utils.LogDebug("go fund", "", "", nil)
+		for {
+			utils.LogDebug("Starting for loop", "", "", nil)
```
Comment (YourTechBud) [inline runner/utils/driver/docker/docker.go:173]: Remove this

File: runner/utils/driver/docker/docker.go [hunk 17]
```diff
@@ -124,6 +126,68 @@ func (d *Docker) ApplyService(ctx context.Context, service *model.Service) error
...
+			str, err := rd.ReadString('\n')
+			if err != nil {
+				if err == io.EOF && !isFollow {
+					utils.LogDebug("End of file reached for logs", "docker", "GetLogs", map[string]interface{}{})
```
Comment (YourTechBud) [inline runner/utils/driver/docker/docker.go:177]:
```suggestion
					utils.LogDebug("End of file reached for logs", "docker", "get-logs", map[string]interface{}{})
```

File: runner/utils/driver/docker/docker.go [hunk 18]
```diff
@@ -124,6 +126,68 @@ func (d *Docker) ApplyService(ctx context.Context, service *model.Service) error
...
+					return
+				}
+				_ = utils.LogError("Unable to read logs from container", "docker", "GetLogs", err)
```
Comment (YourTechBud) [inline runner/utils/driver/docker/docker.go:180]:
```suggestion
				_ = utils.LogError("Unable to read logs from container", "docker", "get-logs", err)
```

File: runner/utils/driver/istio/logs.go [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+		Timestamps: true,
+	})
+
+	b, err := req.Stream(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	pipeReader, pipeWriter := io.Pipe()
+	utils.LogDebug("Sending logs to client", "docker", "GetLogs", map[string]interface{}{})
```
Comment (YourTechBud) [inline runner/utils/driver/istio/logs.go:37]:
```suggestion
	utils.LogDebug("Sending logs to client", "k8s", "get-logs", map[string]interface{}{})
```

File: runner/utils/driver/istio/logs.go [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+)
+
+// GetLogs get logs of specified services
+func (i *Istio) GetLogs(ctx context.Context, isFollow bool, projectID, taskID, replica string) (io.ReadCloser, error) {
+	if taskID == "" {
+		arr := strings.Split(replica, "--")
+		if len(arr) < 2 {
+			return nil, utils.LogError("Invalid replica id", "docker", "get-logs", nil)
+		}
+		taskID = arr[0]
```
Comment (YourTechBud) [inline runner/utils/driver/istio/logs.go:22]: This seems to be wrong. Did you test it?

File: runner/utils/driver/istio/logs.go [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+	go func() {
+		defer utils.CloseTheCloser(b)
+		defer utils.CloseTheCloser(pipeWriter)
+		// read logs
+		rd := bufio.NewReader(b)
+		for {
+			str, err := rd.ReadString('\n')
+			if err != nil {
+				if err == io.EOF && !isFollow {
+					utils.LogDebug("End of file reached for logs", "docker", "GetLogs", map[string]interface{}{})
```
Comment (YourTechBud) [inline runner/utils/driver/istio/logs.go:47]:
```suggestion
					utils.LogDebug("End of file reached for logs", "k8s", "get-logs", map[string]interface{}{})
```

File: runner/utils/driver/istio/logs.go [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+					return
+				}
+				_ = utils.LogError("Unable to read logs from container", "docker", "GetLogs", err)
```
Comment (YourTechBud) [inline runner/utils/driver/istio/logs.go:50]:
```suggestion
				_ = utils.LogError("Unable to read logs from container", "k8s", "get-logs", err)
```

File: runner/utils/driver/istio/logs.go (hunk 20 above)
Comment (sharadregoti) [reply runner/utils/driver/istio/logs.go:22]: yes

**Pull request merged**